JAVA-5479: Add GitHub Actions based release automation

alcaeus opened the pull request:
This PR adds a GitHub Actions Workflow to create releases. The process is as follows:

* Users with write access trigger a new workflow run for the "Release new version" workflow, entering the version to be released
* The automation does some sanity checks on the branch name:
** Release `<major>.<minor>.0` must be released from the maintenance branch (i.e. `<major>.<minor>.x`) or the `master` branch. If the `master` branch is selected, the automation will attempt to create a new maintenance branch
** Release `<major>.<minor>.<patch>` must be released from the maintenance branch (i.e. `<major>.<minor>.x`) - releasing from any other branches is not allowed
* The automation then updates the build version to `<major>.<minor>.<patch>` in `build.gradle`, commits this change and creates a tag
* After tagging, the version in `build.gradle` is updated to `<major>.<minor>.<patch+1>-SNAPSHOT` and again committed
* The maintenance branch and tag are pushed to the repository
* A draft release with generated release notes is created

I've tested the flow in my fork where I've confirmed it works. However, a couple of changes may be appropriate still:
- The automation currently uses the GitHub token of the user triggering the release. Since commits are being pushed directly, branch protection rules may interfere with the release process. In this case, we should add the `mongodb-dbx-release-bot` app to the repository and generate a token for the app. We can then allow the app to bypass branch protections
- If not all users with write access should have permission to create releases, we can set up an environment in GitHub settings and require that an authorised user approves the workflow run
- The release commits and tag are not signed; this could be added using the [git-sign action](https://github.com/mongodb-labs/drivers-github-tools/tree/main?tab=readme-ov-file#git-sign)
- Other release requirements related to the SSDLC effort (e.g. generating a SAST report, uploading SBOM, etc.) are not added yet and can be integrated based on the shared driver's GitHub Actions once they are ready.

Any additional build tasks (e.g. creating a package to upload to a package manager) can be done in a separate workflow that runs when a tag is pushed.

jyemin commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,105 @@
+          git config user.email "${GIT_AUTHOR_EMAIL}"
+
+      # This step bumps version numbers in build.gradle and creates git artifacts for the release
+      # TODO: this could leverage the git-sign action with some minor changes
```
Let's defer.

alcaeus replied:
Removed the TODO for now.

jyemin commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,105 @@
+
+env:
+  GH_TOKEN: ${{ github.token }}
+  # TODO: Adding the mongodb-dbx-release-automation app to the repository will allow fetching a one-time token and pushing
```
This seems like a necessary thing to enable in the initial version, as we do have branch protection rules enabled.

alcaeus replied:
Correct. In that case, you should follow [these instructions (internal link)](https://wiki.corp.mongodb.com/display/SYSENG/GitHub#GitHub-RequestingaGitHubApplication) to request the `mongodb-dbx-release-bot` application to be added to all relevant repositories.

jyemin replied:
Requested

rozza replied:
🍿 Is this the last blocker?

alcaeus replied:
It is - just waiting for our IT to enable the app here, then we're good to go :)

jyemin commented on .github/workflows/bump-and-tag.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+set -e
+
+# TODO: should this always assume that release X.Y.Z will always be created from X.Y.Z-SNAPSHOT"?
```
I believe our current scripts assume this.  Seems safe.

jyemin commented:
Is there anything the blocks taking this out of Draft?

alcaeus commented:
> Is there anything the blocks taking this out of Draft?

Not really. I've rebased on top of master and added the release environment, which allows us to better protect the necessary credentials and also allows for adding additional checks (such as directly restricting which branches can be released).
Once the app has been added to the repository, we're good to start testing.

jyemin reviewed (approved):
LGTM.  Thank you!
